File: solutions/longestCommonSubsequence.js
/**
 * Given two arrays, return the length of the longest subsequence
 * they have in common. The longest common subsequence is often
 * abbreviated LCS.
 *
 * @example
 * longestCommonSubsequence(
 *  ['A', 'B', 'C', 'B', 'E', 'C', 'A'],
 *  ['B', 'D', 'C', 'A', 'E', 'A']
 * ); // => returns 4 because ['B', 'C', 'E', 'A'] is the LCS
 */
function longestCommonSubsequence(left, right) {
  if (left.length === 0 || right.length === 0) {
    return 0;
  }

  let lastLeft = left.at(-1);
  let lastRight = right.at(-1);

  if (lastLeft === lastRight) {
    return 1 + longestCommonSubsequence(left.slice(0, -1), right.slice(0, -1));
  } else {
    return Math.max(
      longestCommonSubsequence(left, right.slice(0, -1)),
      longestCommonSubsequence(left.slice(0, -1), right),
    );
  }
}

// Memoize each (n,m) pair and this works
function longestCommonSubsequenceRef(left, right, n = left.length, m = right.length) {
  if (n === 0 || m === 0) {
    return 0;
  }

  let lastLeft = left.at(n - 1);
  let lastRight = right.at(m - 1);

  if (lastLeft === lastRight) {
    return 1 + longestCommonSubsequenceRef(left, right, n - 1, m - 1);
  } else {
    return Math.max(
      longestCommonSubsequenceRef(left, right, n, m - 1),
      longestCommonSubsequenceRef(left, right, n - 1, m),
    );
  }
}

function longestCommonSubsequenceBottomUp(left, right) {
  // Create an (n+1)-by-(m+1) array
  // The +1 is there to account for the empty subsequence.
  let LCS = Array(left.length + 1).fill().map(() => Array(right.length + 1).fill(0));

  for (let n = 0; n <= left.length; n++) {
    for (let m = 0; m <= right.length; m++) {
      if (n === 0 || m === 0) {
        LCS[n][m] = 0;
        continue;
      }

      let lastLeft = left.at(n - 1);
      let lastRight = right.at(m - 1);

      if (lastLeft === lastRight) {
        LCS[n][m] = 1 + LCS[n - 1][m - 1];
      } else {
        LCS[n][m] = Math.max(
          LCS[n][m - 1],
          LCS[n - 1][m],
        );
      }
    }
  }

  return LCS[left.length][right.length];
}

if (require.main === module) {
  console.log('Running sanity checks:');

  // Add some sanity checks here
}

module.exports = {
  longestCommonSubsequence,
  longestCommonSubsequenceRef,
  longestCommonSubsequenceBottomUp,
};

File: python/solutions/longest_common_subsequence.py
"""
Given two sequences, return the length of the longest subsequence
they have in common. The longest common subsequence is often
abbreviated LCS.

Example:
longest_common_subsequence(
    ['A', 'B', 'C', 'B', 'E', 'C', 'A'],
    ['B', 'D', 'C', 'A', 'E', 'A']
) # => returns 4 because ['B', 'C', 'E', 'A'] is the LCS
"""

def longest_common_subsequence(left, right):
    if len(left) == 0 or len(right) == 0:
        return 0

    last_left = left[-1]
    last_right = right[-1]

    if last_left == last_right:
        return 1 + longest_common_subsequence(left[:-1], right[:-1])
    else:
        return max(
            longest_common_subsequence(left, right[:-1]),
            longest_common_subsequence(left[:-1], right)
        )

def longest_common_subsequence_ref(left, right, n=None, m=None):
    if n is None:
        n = len(left)
    if m is None:
        m = len(right)

    if n == 0 or m == 0:
        return 0

    last_left = left[n - 1]
    last_right = right[m - 1]

    if last_left == last_right:
        return 1 + longest_common_subsequence_ref(left, right, n - 1, m - 1)
    else:
        return max(
            longest_common_subsequence_ref(left, right, n, m - 1),
            longest_common_subsequence_ref(left, right, n - 1, m)
        )

def longest_common_subsequence_bottom_up(left, right):
    # Create an (n+1)-by-(m+1) array
    # The +1 is there to account for the empty subsequence.
    LCS = [[0 for _ in range(len(right) + 1)] for _ in range(len(left) + 1)]

    for n in range(len(left) + 1):
        for m in range(len(right) + 1):
            if n == 0 or m == 0:
                LCS[n][m] = 0
                continue

            last_left = left[n - 1]
            last_right = right[m - 1]

            if last_left == last_right:
                LCS[n][m] = 1 + LCS[n - 1][m - 1]
            else:
                LCS[n][m] = max(
                    LCS[n][m - 1],
                    LCS[n - 1][m]
                )

    return LCS[len(left)][len(right)]

if __name__ == "__main__":
    print('Running sanity checks:')
    
    # Add some sanity checks here
    test_left = ['A', 'B', 'C', 'B', 'E', 'C', 'A']
    test_right = ['B', 'D', 'C', 'A', 'E', 'A']
    
    assert longest_common_subsequence(test_left, test_right) == 4
    assert longest_common_subsequence_ref(test_left, test_right) == 4
    assert longest_common_subsequence_bottom_up(test_left, test_right) == 4
    
    print("All sanity checks passed!")
